### PR TITLE
Use PID to assign rainbow colors to console output prefixes.

### PR DESCRIPTION
**Pull request by ethanabrooks**
## Why are these changes needed?

When I use `tune.run`, different processes are distinguished by the number corresponding to their pid:
image

This pull request implements color output to distinguish between processes similar to docker-compose:
image

## Related issue number 
https://github.com/ray-project/ray/issues/26226

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [x] This PR is not tested :(

### CODE REVIEW

**Comment by rkooo567** [comment]
There are lots of test failures. Can you take a look?

File: python/ray/_private/worker.py [hunk 1]
```diff
@@ -1641,14 +1641,29 @@ def color_for(data: Dict[str, str], line: str) -> str:
             data.get("pid") == "raylet"
             and ray_constants.LOG_PREFIX_INFO_MESSAGE not in line
         ):
-            return colorama.Fore.YELLOW
+            return colorama.Fore.RED
```
Comment (rkooo567) [inline python/ray/_private/worker.py:1644]: Can you keep the same color for this? What was the motivation for the color change?

**Comment by ethanabrooks** [comment]
Yes certainly. I can have a look into this later today.

File: python/ray/_private/worker.py (hunk 1 above)
Comment (ethanabrooks) [reply python/ray/_private/worker.py]: Yes I reverted that. I had thought to mention it -- just seems like if we are going to use the color red, we should use it for errors -- but yellow works fine too.

File: python/ray/_private/worker.py [hunk 2]
```diff
@@ -1648,7 +1648,22 @@ def color_for(data: Dict[str, str], line: str) -> str:
             else:
                 return colorama.Style.BRIGHT + colorama.Fore.CYAN
         else:
-            return colorama.Fore.CYAN
+            colors = [
+                colorama.Fore.BLUE,
+                colorama.Fore.CYAN,
```
Comment (rkooo567) [inline python/ray/_private/worker.py:1653]: Can you remove CYAN since it can be used for the autoscaler log?

**Review by rkooo567 — APPROVED** [review]
LGTM! One last feedback. Let me also try running this locally.

**Comment by rkooo567** [comment]
Btw, you need to use `return colors[int(data.get("pid", 0)) % len(colors)]` to fix the bug.

Also I found some of colors are too light for the white terminal. Do you know how other tools choose colors based on the terminal color?

**Comment by rkooo567** [comment]
Example:

File: python/ray/_private/worker.py (hunk 2 above)
Comment (ethanabrooks) [reply python/ray/_private/worker.py]: This is done.

**Comment by ethanabrooks** [comment]
> Also I found some of colors are too light for the white terminal. Do you know how other tools choose colors based on the terminal color? I feel like black terminals probably can have the same issue (maybe avoiding light color will fix the issue?)

How does this look?
image
Maybe we want to eliminate `WHITE`, `LIGHTCYAN_EX` or `LIGHTGREEN_EX`?

**Comment by rkooo567** [comment]
@ethanabrooks Yes! That sounds good to me. Can you also show me the output on the black screen?

**Comment by rkooo567** [comment]
There are still some failures such as 

```
E       assert 'Spilled ' in 'Exception in thread ray_print_logs:\nTraceback (most recent call last):\n  File "/opt/miniconda/lib/python3.7/threadi...turn colors[int(data.get("pid", 0)) % len(colors)]\nValueError: invalid literal for int() with base 10: \'raylet\'\n\n'
```

! Thanks again for addressing the comment :)!

**Comment by ethanabrooks** [comment]
Looks like "pid" is not always an int. Just pushed a simple fix.

**Comment by ethanabrooks** [comment]
> @ethanabrooks Yes! That sounds good to me. Can you also show me the output on the black screen?

image

**Comment by rkooo567** [comment]
let's also remove blue. Can you also merge the latest master? I think we are almost ready to go :)!

**Comment by ethanabrooks** [comment]
Blue is removed.

**Review by ericl — CHANGES_REQUESTED** [review]
Hey guys, I really don't think we should do this, or at least not by default. How about making it a pluggable function instead?

Turning the default output for users into a rainbow is not really a clean experience.

**Comment by rkooo567** [comment]
Sounds good for me the pluggable function. In this case, we might need API approval for that.

OOC, why is it not a clean experience? Given the color is not badly compatible to the terminal background, isn't it a strictly a better experience (since it will make it easier to distinguish output from different pids)

**Comment by scottsun94** [comment]
I tested the ANSI normal and bright colors against common background colors (caveat: this is tested on Mac terminal and how Mac maps the ANSI colors to actual hex values may be different from other terminals):
Screen Shot 2022-08-01 at 5 34 20 PM

To me, we should remove blue, white, light/bright cyan, light/bright green (too light on one of the backgrounds) and red (this should be reserved for errors).

Regardless whether this will be a plugin or not, the final list could be. 
- Magenta
- Green
- Cyan
- Lightblack
- Lightblue
- Lightmagenta

Thanks for sharing the screenshots of the colors in your terminal! Color is a tricky thing!

**Comment by ethanabrooks** [comment]
Not a problem. This seems like a good list, although yellow is currently being used for errors if I'm not mistaken.

**Comment by scottsun94** [comment]
@ethanabrooks You are right. We use yellow for errors currently. I'm not sure if this is the best way to go.
Regardless, red tends to people think of errors since most of other systems use red for errors. Even Ray doesn't end up with using red for errors in the future, it's probably still good to reserve it. (Would like to hear your thoughts too. Yellow vs red for errors.)

**Comment by ethanabrooks** [comment]
Well, my initial take was that red should be the error color and that would allow us to add yellow back into the PID colors, although we probably have enough with the six you proposed.

**Comment by scottsun94** [comment]
To me, red should be reserved for sure. Whether to reserve yellow for warning is more debatable. Anyway, let's probably start with the 6 colors for now and iterate based on more feedback from everyone else.

**Comment by ethanabrooks** [comment]
Before I push a revision, I think there was a reserved use for Cyan.

**Comment by scottsun94** [comment]
I'm not sure if Cyan is reserved for any use. @rkooo567 Do you know?

**Comment by rkooo567** [comment]
Looks like CYAN is just the default color (default log color), meaning it is not reserved for anything rn

**Comment by ethanabrooks** [comment]
Hi, I just added cyan back in. Are there any other changes I can make here?

**Comment by ericl** [comment]
Regarding making this a pluggable function, the easiest way is to have it as an env var, and we could also include a default, e.g., RAY_LOG_FORMATTER_HOOK=ray.util.color_log_formattrt.

This could be documented under the Ray logging section in the docs as a tip.

**Comment by ethanabrooks** [comment]
Sounds good. Happy to implement.

**Comment by ethanabrooks** [comment]
I just added an environment variable toggle `RAY_COLOR_PID`. 

@ericl I didn't understand your `RAY_LOG_FORMATTER_HOOK=ray.util.color_log_formatter` example. If you provide more details I can look into implementing.

**Review by ericl — COMMENTED** [review]
The flag as implemented is good enough for me! How about also adding a screenshot / explanation on how to enable it to this page? (logging.rst) https://docs.ray.io/en/master/ray-observability/ray-logging.html

**Comment by ethanabrooks** [comment]
Good suggestion. I can work on that.

**Comment by scottsun94** [comment]
@ethanabrooks Have you got a chance to add the related documentation? Looking forward to this improvement.

**Comment by ethanabrooks** [comment]
Thanks for the reminder. I should have time this weekend.

**Comment by rkooo567** [comment]
Lmk when it is ready to review again :) !

**Comment by ethanabrooks** [comment]
Yes I will. My apologies, you are catching me in a bit of a busy period. I may have to defer this to the second half of the month (deadline on the 14th).

**Comment by ethanabrooks** [comment]
Still working on this!

**Comment by rkooo567** [comment]
unstale

**Comment by scottsun94** [comment]
> Running into an issue with building locally:
> 
> ```
> >>> import ray
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
>   File "/home/ethanbro/ray-screenshot/venv/lib/python3.6/site-packages/ray/__init__.py", line 109, in <module>
>     import ray._raylet  # noqa: E402
>   File "python/ray/_raylet.pyx", line 118, in init ray._raylet
>     RayActorError,
>   File "/home/ethanbro/ray-screenshot/venv/lib/python3.6/site-packages/ray/exceptions.py", line 17, in <module>
>     from ray.util.annotations import DeveloperAPI, PublicAPI
>   File "/home/ethanbro/ray-screenshot/venv/lib/python3.6/site-packages/ray/util/__init__.py", line 5, in <module>
>     from ray._private.services import get_node_ip_address
>   File "/home/ethanbro/ray-screenshot/venv/lib/python3.6/site-packages/ray/_private/services.py", line 28, in <module>
>     from ray._private.gcs_utils import GcsClient
>   File "/home/ethanbro/ray-screenshot/venv/lib/python3.6/site-packages/ray/_private/gcs_utils.py", line 13, in <module>
>     from ray.core.generated.gcs_pb2 import (
> ImportError: cannot import name 'JobConfig'
> ```
> 
> Here is a gist of ray/core/generated/gcs_pb2.py: https://gist.github.com/ethanabrooks/cc2f647413680efae7a016cd126bbda2
> 
> Using these instructions: https://docs.ray.io/en/master/ray-contribute/development.html#prepare-the-python-environment

Are you still seeing this issue?

**Comment by ethanabrooks** [comment]
I am.

**Comment by rkooo567** [comment]
Try `bazel build //:ray_pkg`?

File: doc/source/ray-observability/ray-logging.rst [hunk 3]
```diff
@@ -70,6 +70,11 @@ This produces the following output:
     (MyActor(index=2) pid=482120) hello there
     (MyActor(index=1) pid=482119) hello there
 
+Coloring Actor logs prefixes
```
Comment (angelinalg) [inline doc/source/ray-observability/ray-logging.rst:73]:
```suggestion
Coloring Actor log prefixes
```

Comment (angelinalg) [reply doc/source/ray-observability/ray-logging.rst:73]: Hi, I'm offering some unsolicited editing to try to be helpful. Not sure if this changes the meaning, but typically, when a noun is used as an adjective, it is singular, not plural.

File: doc/source/ray-observability/ray-logging.rst [hunk 4]
```diff
@@ -70,6 +70,11 @@ This produces the following output:
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+By default Ray prints Actor logs prefixes in light blue:
```
Comment (angelinalg) [inline doc/source/ray-observability/ray-logging.rst:75]:
```suggestion
By default, Ray prints Actor log prefixes in light blue:
```

File: doc/source/ray-observability/ray-logging.rst [hunk 5]
```diff
@@ -70,6 +70,11 @@ This produces the following output:
...
+Users may instead activate colored prefixes by setting the environment variable ``RAY_COLOR_PREFIX=1``:
```
Comment (angelinalg) [inline doc/source/ray-observability/ray-logging.rst:76]:
```suggestion
Alternatively, you can activate colored prefixes by setting the environment variable ``RAY_COLOR_PREFIX=1``:
```

File: doc/source/ray-observability/ray-logging.rst [hunk 6]
```diff
@@ -70,6 +70,11 @@ This produces the following output:
...
+
 How to set up loggers
 ~~~~~~~~~~~~~~~~~~~~~
 When using ray, all of the tasks and actors are executed remotely in Ray's worker processes.
```
Comment (angelinalg) [inline doc/source/ray-observability/ray-logging.rst:80]:
```suggestion
When using Ray, all tasks and actors are executed remotely in Ray's worker processes. 
```

**Comment by ethanabrooks** [comment]
> 

❯ bazel build //:ray_pkg
Extracting Bazel installation...
Starting local Bazel server and connecting to it...
DEBUG: /home/ethanbro/ray/bazel/ray_deps_setup.bzl:67:14: No implicit mirrors used because urls were explicitly provided
ERROR: /home/ethanbro/.cache/bazel/_bazel_ethanbro/b5f785192cb7aa69e364317960df121d/external/bazel_tools/platforms/BUILD:89:6: in alias rule @bazel_tools//platforms:windows: Constraints from @bazel_tools//platforms have been removed. Please use constraints from @platforms repository embedded in Bazel, or preferably declare dependency on https://github.com/bazelbuild/platforms. See https://github.com/bazelbuild/bazel/issues/8622 for details.
ERROR: /home/ethanbro/.cache/bazel/_bazel_ethanbro/b5f785192cb7aa69e364317960df121d/external/bazel_tools/platforms/BUILD:89:6: Analysis of target '@bazel_tools//platforms:windows' failed
ERROR: /home/ethanbro/.cache/bazel/_bazel_ethanbro/b5f785192cb7aa69e364317960df121d/external/com_google_protobuf/BUILD:513:10: While resolving toolchains for target @com_google_protobuf//:protoc: invalid registered toolchain '@bazel_skylib//toolchains/unittest:cmd_toolchain':
ERROR: Analysis of target '//:ray_pkg' failed; build aborted:
INFO: Elapsed time: 37.263s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (45 packages loaded, 1145 targets configured)

**Comment by ethanabrooks** [comment]
> 



> Try `bazel build //:ray_pkg`?

```
❯ bazel build //:ray_pkg
Extracting Bazel installation...
Starting local Bazel server and connecting to it...
DEBUG: /home/ethanbro/ray/bazel/ray_deps_setup.bzl:67:14: No implicit mirrors used because urls were explicitly provided
ERROR: /home/ethanbro/.cache/bazel/_bazel_ethanbro/b5f785192cb7aa69e364317960df121d/external/bazel_tools/platforms/BUILD:89:6: in alias rule @bazel_tools//platforms:windows: Constraints from @bazel_tools//platforms have been removed. Please use constraints from @platforms repository embedded in Bazel, or preferably declare dependency on https://github.com/bazelbuild/platforms. See https://github.com/bazelbuild/bazel/issues/8622 for details.
ERROR: /home/ethanbro/.cache/bazel/_bazel_ethanbro/b5f785192cb7aa69e364317960df121d/external/bazel_tools/platforms/BUILD:89:6: Analysis of target '@bazel_tools//platforms:windows' failed
ERROR: /home/ethanbro/.cache/bazel/_bazel_ethanbro/b5f785192cb7aa69e364317960df121d/external/com_google_protobuf/BUILD:513:10: While resolving toolchains for target @com_google_protobuf//:protoc: invalid registered toolchain '@bazel_skylib//toolchains/unittest:cmd_toolchain':
ERROR: Analysis of target '//:ray_pkg' failed; build aborted:
INFO: Elapsed time: 37.263s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (45 packages loaded, 1145 targets configured)
```

**Review by rkooo567 — APPROVED** [review]

**Comment by rkooo567** [comment]
I think you should try bazel clean --expunge first!

**Comment by rkooo567** [comment]
Also can you try merging the latest master?

**Comment by ethanabrooks** [comment]
Hi all, sorry for the delay in working on this. I have merged in the latest changes from master and added a screenshot to the documentation. Please do let me know how it looks.

**Comment by rkooo567** [comment]
Screen Shot 2023-03-22 at 12 53 27 AM

I belive the image is wrong?

**Comment by ethanabrooks** [comment]
Apologies! Grabbed the wrong screenshot :(

**Review by rkooo567 — APPROVED** [review]

**Review by ericl — APPROVED** [review]

**Comment by rkooo567** [comment]
Thanks for the contribution!

**Comment by ethanabrooks** [comment]
My pleasure! Glad e finally got this across the finish line :)

**Comment by scottsun94** [comment]
Screen Shot 2023-05-02 at 8 52 30 PM

@ethanabrooks It seems that the screenshot is broken on the master doc. Do you know why? https://docs.ray.io/en/master/ray-observability/ray-logging.html

**Comment by ethanabrooks** [comment]
@scottsun94 looks like there was an issue with the link I used.  Just submitted a request to fix: https://github.com/ray-project/ray/pull/34998/files

**Pull request merged**